File: vnproglib-1.1/net/src/Protocol/Uart/PacketFinder.cs
using System;
using System.Collections.Generic;

namespace VectorNav.Protocol.Uart
{

/// <summary>
/// Helper class for finding packets in raw incoming data from a VectorNav
/// sensor.
/// </summary>
/// <remarks>
/// Internally, the PacketFinder keeps track of a running data index which
/// keeps a running count of the bytes that are processed by the class. This is
/// useful for users who wish to keep track of where packets where found in the
/// incoming raw data stream. When the PacketFinder receives its first byte
/// from the user, this is given the index of 0 for the running index and
/// incremented for each byte received.
/// </remarks>
public class PacketFinder
{
	private const int DefaultReceiveBufferSize = 512;
	private const int MaximumSizeForAsciiPacket = 256;
	private const int MaximumSizeExpectedForBinaryPacket = 256;

	#region Types

	private class AsciiTracker
	{
		public bool CurrentlyBuildingAsciiPacket;
		public int PossibleStartOfPacketIndex;
		public bool AsciiEndChar1Found;
		public int RunningDataIndexOfStart;

		public void Reset()
		{
			CurrentlyBuildingAsciiPacket = false;
			PossibleStartOfPacketIndex = 0;
			AsciiEndChar1Found = false;
			RunningDataIndexOfStart = 0;
		}
	}

	private class BinaryTracker
	{
		public int PossibleStartIndex;
		public bool GroupsPresentFound;
		public byte GroupsPresent;
		public byte NumOfBytesRemainingToHaveAllGroupFields;
		public int NumOfBytesRemainingForCompletePacket;
		public bool StartFoundInProvidedDataBuffer = true;
		public readonly int RunningDataIndexOfStart;

		public BinaryTracker(int possibleStartIndex, int runningDataIndex)
		{
			PossibleStartIndex = possibleStartIndex;
			RunningDataIndexOfStart = runningDataIndex;
		}

		/*
		bool operator==(const BinaryTracker &rhs)
		{
			return
				possibleStartIndex == rhs.possibleStartIndex &&
				groupsPresentFound == rhs.groupsPresentFound &&
				groupsPresent == rhs.groupsPresent &&
				numOfBytesRemainingToHaveAllGroupFields == rhs.numOfBytesRemainingToHaveAllGroupFields &&
				numOfBytesRemainingForCompletePacket == rhs.numOfBytesRemainingForCompletePacket;
		}
		 */
	}

	#endregion

	#region Events

	/// <summary>
	/// Allows registering for notification of when a valid packet if found.
	/// </summary>
	public event EventHandler<PacketFoundEventArgs> ValidPacketFound;

	#endregion

	#region Constructors

	/// <summary>
	/// Creates a new PacketFinder with internal buffers to store incoming
	/// bytes and alert when valid packets are received.
	/// </summary>
	public PacketFinder() : this(DefaultReceiveBufferSize)
	{
	}

	/// <summary>
	/// Creates a new PacketFinder with an internal buffer the size specified.
	/// </summary>
	/// <param name="receiveBufferSize">
	/// The number of bytes to make the internal buffer.
	/// </param>
	public PacketFinder(int receiveBufferSize)
	{
		_buffer = new byte[receiveBufferSize];
	}

	#endregion

	#region Methods

	private void ResetTracking()
	{
		_asciiOnDeck.Reset();
		_binaryOnDeck.Clear();
		_bufferAppendLocation = 0;
	}

	private void OnValidPacketFoundEvent(Packet foundPacket, int runningIndexOfPacketStart)
	{
		if (ValidPacketFound != null)
			ValidPacketFound(this, new PacketFoundEventArgs(foundPacket, runningIndexOfPacketStart));
	}

	/// <summary>
	/// Adds new data to the internal buffers and processes the received data
	/// to determine if any new received packets are available.
	/// </summary>
	/// <param name="data">
	/// The data buffer containing the received data.
	/// </param>
	/// <param name="index">
	/// The start index of the received data.
	/// </param>
	/// <param name="length">
	/// The number of bytes received.
	/// </param>
	public void ProcessReceivedData(byte[] data, int index, int length)
	{
		var asciiStartFoundInProvidedBuffer = false;

		// Assume that since the _runningDataIndex is unsigned, any overflows
		// will naturally go to zero, which is the behavior that we want.
		for (var i = 0; i < length; i++, _runningDataIndex++)
		{
			var cur = data[index + i];

			if (cur == Packet.AsciiStartChar)
			{
				_asciiOnDeck.Reset();
				_asciiOnDeck.CurrentlyBuildingAsciiPacket = true;
				_asciiOnDeck.PossibleStartOfPacketIndex = i;
				_asciiOnDeck.RunningDataIndexOfStart = _runningDataIndex;

				asciiStartFoundInProvidedBuffer = true;
			}
			else if (_asciiOnDeck.CurrentlyBuildingAsciiPacket && cur == Packet.AsciiEndChar1)
			{
				_asciiOnDeck.AsciiEndChar1Found = true;
			}
			else if (_asciiOnDeck.AsciiEndChar1Found)
			{
				if (cur == Packet.AsciiEndChar2)
				{
					// We have a possible data packet.
					byte[] packetBuffer = null;
					var startIndex = 0;
					var packetLength = 0;

					if (asciiStartFoundInProvidedBuffer)
					{
						// All of the packet data was in the provided buffer so
						// we don't need to any copying.
						packetBuffer = data;
						startIndex = index + _asciiOnDeck.PossibleStartOfPacketIndex;
						packetLength = i - _asciiOnDeck.PossibleStartOfPacketIndex + 1;
					}
					else
					{
						// The packet is split between the internal and the
						// provided data buffers. We need to copy the data over
						// before futher processing.

						if (_bufferAppendLocation + i < _buffer.Length)
						{
							Buffer.BlockCopy(data, index, _buffer, _bufferAppendLocation, i + 1);

							packetBuffer = _buffer;
							startIndex = _asciiOnDeck.PossibleStartOfPacketIndex;
							packetLength = _bufferAppendLocation + i + 1 - _asciiOnDeck.PossibleStartOfPacketIndex;
						}
						else
						{
							// We are about to overflow our buffer. Just fall
							// through to reset tracking.
						}
					}

					var p = new Packet(packetBuffer, startIndex, packetLength);

					if (p.IsValid)
						OnValidPacketFoundEvent(p, _asciiOnDeck.RunningDataIndexOfStart);
				}

				// Either this is an invalid packet or was a packet that was processed.
				if (_binaryOnDeck.Count == 0)
					ResetTracking();
				else
					_asciiOnDeck.Reset();

				asciiStartFoundInProvidedBuffer = false;
			}
			else if (i + 1 > MaximumSizeForAsciiPacket)
			{
				// This must not be a valid ASCII packet.
				if (_binaryOnDeck.Count == 0)
					ResetTracking();
				else
					_asciiOnDeck.Reset();

				asciiStartFoundInProvidedBuffer = false;
			}

			// Update all of our binary packets on deck.
			var invalidPackets = new List<BinaryTracker>();
			foreach (var bp in _binaryOnDeck)
			{
				if (!bp.GroupsPresentFound)
				{
					// This byte must be the groups present.
					bp.GroupsPresentFound = true;
					bp.GroupsPresent = cur;
					bp.NumOfBytesRemainingToHaveAllGroupFields = (byte) (2 * Data.Util.CountSetBits(cur));

					continue;
				}

				if (bp.NumOfBytesRemainingToHaveAllGroupFields != 0)
				{
					// We found another byte belonging to this possible binary packet.
					bp.NumOfBytesRemainingToHaveAllGroupFields--;

					if (bp.NumOfBytesRemainingToHaveAllGroupFields == 0)
					{
						// We have all of the group fields now.

						int remainingBytesForCompletePacket;

						if (bp.StartFoundInProvidedDataBuffer)
						{
							var headerLength = i - bp.PossibleStartIndex + 1;
							remainingBytesForCompletePacket = Packet.ComputeBinaryPacketLength(data, index + bp.PossibleStartIndex) - headerLength;
						}
						else
						{
							// Not all of the packet's group is inside the caller's provided buffer.

							// Temporarily copy the rest of the packet to the receive buffer
							// for computing the size of the packet.

							var numOfBytesToCopyIntoReceiveBuffer = i + 1;
							var headerLength = _bufferAppendLocation - bp.PossibleStartIndex + numOfBytesToCopyIntoReceiveBuffer;

							if (_bufferAppendLocation + numOfBytesToCopyIntoReceiveBuffer < _buffer.Length)
							{
								Buffer.BlockCopy(data, index, _buffer, _bufferAppendLocation, numOfBytesToCopyIntoReceiveBuffer);

								remainingBytesForCompletePacket = Packet.ComputeBinaryPacketLength(_buffer, bp.PossibleStartIndex) - headerLength;
							}
							else
							{
								// About to overrun our receive buffer!
								invalidPackets.Add(bp);

								continue;
							}
						}

						if (remainingBytesForCompletePacket > MaximumSizeExpectedForBinaryPacket)
							// Must be a bad possible binary packet.
							invalidPackets.Add(bp);
						else
							bp.NumOfBytesRemainingForCompletePacket = remainingBytesForCompletePacket;
					}

					continue;
				}

				// We are currently collecting data for our packet.

				bp.NumOfBytesRemainingForCompletePacket--;

				if (bp.NumOfBytesRemainingForCompletePacket == 0)
				{
					// We have a possible binary packet!
					byte[] packetBuffer;
					int startIndex;
					int packetLength;

					if (bp.StartFoundInProvidedDataBuffer)
					{
						// The binary packet exists completely in the user's provided buffer.
						packetBuffer = data;
						startIndex = bp.PossibleStartIndex;
						packetLength = i - bp.PossibleStartIndex + 1;
					}
					else
					{
						// The packet is split between our receive buffer and the user's buffer.
						var numOfBytesToCopyIntoReceiveBuffer = i + 1;

						if (_bufferAppendLocation + numOfBytesToCopyIntoReceiveBuffer < _buffer.Length)
						{
							Buffer.BlockCopy(data, index, _buffer, _bufferAppendLocation, numOfBytesToCopyIntoReceiveBuffer);

							packetBuffer = _buffer;
							startIndex = bp.PossibleStartIndex;
							packetLength = _bufferAppendLocation - bp.PossibleStartIndex + i + 1;
						}
						else
						{
							// About to overrun our receive buffer!
							invalidPackets.Add(bp);

							continue;
						}
					}

					var p = new Packet(packetBuffer, startIndex, packetLength);

					if (!p.IsValid)
					{
						// Invalid packet!
						invalidPackets.Add(bp);
					}
					else
					{
						// We have a valid binary packet!!!.
						OnValidPacketFoundEvent(p, bp.RunningDataIndexOfStart);

						invalidPackets.Clear();
						ResetTracking();
						asciiStartFoundInProvidedBuffer = false;

						break;
					}
				}
			}

			// Remove any invalid packets.
			foreach (var ip in invalidPackets)
				_binaryOnDeck.Remove(ip);

			if (_binaryOnDeck.Count == 0 && !_asciiOnDeck.CurrentlyBuildingAsciiPacket)
				_bufferAppendLocation = 0;

			if (cur == Packet.BinaryStartChar)
				// Possible start of a binary packet.
				_binaryOnDeck.Add(new BinaryTracker(i, _runningDataIndex));
		}

		if (_binaryOnDeck.Count == 0 && !_asciiOnDeck.CurrentlyBuildingAsciiPacket)
			// No data to copy over.
			return;

		// Perform any data copying to our internal buffer.
		var binaryDataToCopyOver = false;
		var dataIndexToStartCopyingFrom = 0;
		var binaryDataMoveOverIndexAdjustment = 0;

		if (_binaryOnDeck.Count != 0)
		{
			binaryDataToCopyOver = true;

			if (_binaryOnDeck[0].StartFoundInProvidedDataBuffer)
			{
				dataIndexToStartCopyingFrom = _binaryOnDeck[0].PossibleStartIndex;
				binaryDataMoveOverIndexAdjustment = dataIndexToStartCopyingFrom;
			}
		}

		if (_asciiOnDeck.CurrentlyBuildingAsciiPacket && asciiStartFoundInProvidedBuffer)
		{
			if (_asciiOnDeck.PossibleStartOfPacketIndex < dataIndexToStartCopyingFrom)
			{
				binaryDataMoveOverIndexAdjustment -= binaryDataMoveOverIndexAdjustment - _asciiOnDeck.PossibleStartOfPacketIndex;
				dataIndexToStartCopyingFrom = _asciiOnDeck.PossibleStartOfPacketIndex;
			}
			else if (!binaryDataToCopyOver)
			{
				dataIndexToStartCopyingFrom = _asciiOnDeck.PossibleStartOfPacketIndex;
			}

			// Adjust our ASCII index to be based on the recieve buffer.
			_asciiOnDeck.PossibleStartOfPacketIndex = _bufferAppendLocation + _asciiOnDeck.PossibleStartOfPacketIndex - dataIndexToStartCopyingFrom;
		}

		// Adjust any binary packet indexes we are currently building.
		foreach (var bp in _binaryOnDeck)
		{
			if (bp.StartFoundInProvidedDataBuffer)
			{
				bp.StartFoundInProvidedDataBuffer = false;
				bp.PossibleStartIndex = bp.PossibleStartIndex - binaryDataMoveOverIndexAdjustment + _bufferAppendLocation;
			}
		}

		if (_bufferAppendLocation + length - dataIndexToStartCopyingFrom < _buffer.Length)
		{
			// Safe to copy over the data.

			var numOfBytesToCopyOver = length - dataIndexToStartCopyingFrom;

			Buffer.BlockCopy(data, dataIndexToStartCopyingFrom, _buffer, _bufferAppendLocation, numOfBytesToCopyOver);

			_bufferAppendLocation += numOfBytesToCopyOver;
		}
		else
		{
			// We are about to overflow our buffer.
			ResetTracking();

			_bufferAppendLocation = 0;
		}
	}

	#endregion

	private readonly byte[] _buffer;
	private int _bufferAppendLocation;
	private readonly AsciiTracker _asciiOnDeck = new AsciiTracker();
	private readonly List<BinaryTracker> _binaryOnDeck = new List<BinaryTracker>(); 

	// Used for correlating raw data received with where packets are found.
	private int _runningDataIndex;
}

}

File: vnproglib-1.1.5.0/cpp/src/packetfinder.cpp
#include "vn/packetfinder.h"
#include "vn/utilities.h"
#include "vn/error_detection.h"

#include <queue>
#include <list>
#include <cstring>

#if PYTHON
	#include "boostpython.h"
	namespace bp = boost::python;
#endif

using namespace std;
using namespace vn::xplat;
using namespace vn::data::integrity;

namespace vn {
namespace protocol {
namespace uart {

//char* vnstrtok(char* str, size_t& startIndex);

struct BinaryTracker
{
	size_t possibleStartIndex;
	bool groupsPresentFound;
	uint8_t groupsPresent;
	uint8_t numOfBytesRemainingToHaveAllGroupFields;
	size_t numOfBytesRemainingForCompletePacket;
	bool startFoundInProvidedDataBuffer;
	size_t runningDataIndexOfStart;
	vn::xplat::TimeStamp timeFound;
	explicit BinaryTracker(size_t possibleStartIndex, size_t runningDataIndex, TimeStamp timeFound_) :
		possibleStartIndex(possibleStartIndex),
		groupsPresentFound(false),
		numOfBytesRemainingToHaveAllGroupFields(0),
		numOfBytesRemainingForCompletePacket(0),
		startFoundInProvidedDataBuffer(true),
		runningDataIndexOfStart(runningDataIndex),
		timeFound(timeFound_)
	{ }
};

bool operator==(const BinaryTracker& lhs, const BinaryTracker& rhs)
{
	return
		lhs.possibleStartIndex == rhs.possibleStartIndex &&
		lhs.groupsPresentFound == rhs.groupsPresentFound &&
		lhs.groupsPresent == rhs.groupsPresent &&
		lhs.numOfBytesRemainingToHaveAllGroupFields == rhs.numOfBytesRemainingToHaveAllGroupFields &&
		lhs.numOfBytesRemainingForCompletePacket == rhs.numOfBytesRemainingForCompletePacket;
}

struct PacketFinder::Impl
{
	static const size_t DefaultReceiveBufferSize = 512;
	static const uint8_t AsciiStartChar = '$';
	static const uint8_t BinaryStartChar = 0xFA;
	static const uint8_t AsciiEndChar1 = '\r';
	static const uint8_t AsciiEndChar2 = '\n';
	static const size_t MaximumSizeExpectedForBinaryPacket = 256;
	static const size_t MaximumSizeForBinaryStartAndAllGroupData = 18;
	static const size_t MaximumSizeForAsciiPacket = 256;

	struct AsciiTracker
	{
		bool currentlyBuildingAsciiPacket;
		size_t possibleStartOfPacketIndex;
		bool asciiEndChar1Found;
		size_t runningDataIndexOfStart;
		TimeStamp timeFound;

		AsciiTracker() :
			currentlyBuildingAsciiPacket(false),
			possibleStartOfPacketIndex(0),
			asciiEndChar1Found(false),
			runningDataIndexOfStart(0)
		{ }

		void reset()
		{
			currentlyBuildingAsciiPacket = false;
			possibleStartOfPacketIndex = 0;
			asciiEndChar1Found = false;
			runningDataIndexOfStart = 0;
			timeFound = TimeStamp();
		}
	};

	PacketFinder* _backReference;
	uint8_t* _buffer;
	const size_t _bufferSize;
	size_t _bufferAppendLocation;
	AsciiTracker _asciiOnDeck;
	list<BinaryTracker> _binaryOnDeck;	// Collection of possible binary packets we are checking.
	size_t _runningDataIndex;			// Used for correlating raw data with where the packet was found for the end user.
	void* _possiblePacketFoundUserData;
	ValidPacketFoundHandler _possiblePacketFoundHandler;
	#if PYTHON
	/*boost::python::object* _pythonPacketFoundHandler;*/
	PyObject* _pythonPacketFoundHandler;
	#endif

	explicit Impl(PacketFinder* backReference) :
		_backReference(backReference),
		_buffer(new uint8_t[DefaultReceiveBufferSize]),
		_bufferSize(DefaultReceiveBufferSize),
		_bufferAppendLocation(0),
		_runningDataIndex(0),
		_possiblePacketFoundUserData(NULL),
		_possiblePacketFoundHandler(NULL)
		#if PYTHON
		,
		_pythonPacketFoundHandler(NULL)
		#endif
	{ }

	Impl(PacketFinder* backReference, size_t internalReceiveBufferSize) :
		_backReference(backReference),
		_buffer(new uint8_t[internalReceiveBufferSize]),
		_bufferSize(internalReceiveBufferSize),
		_bufferAppendLocation(0),
		_runningDataIndex(0),
		_possiblePacketFoundUserData(NULL),
		_possiblePacketFoundHandler(NULL)
	{ }

	~Impl()
	{
		delete [] _buffer;
	}

	void resetTracking()
	{
		_asciiOnDeck.reset();
		_binaryOnDeck.clear();
		_bufferAppendLocation = 0;
	}

	void dataReceived(uint8_t data[], size_t length, TimeStamp timestamp)
	{
		bool asciiStartFoundInProvidedBuffer = false;

		// Assume that since the _runningDataIndex is unsigned, any overflows
		// will naturally go to zero, which is the behavior that we want.
		for (size_t i = 0; i < length; i++, _runningDataIndex++)
		{
			if (data[i] == AsciiStartChar)
			{
				_asciiOnDeck.reset();
				_asciiOnDeck.currentlyBuildingAsciiPacket = true;
				_asciiOnDeck.possibleStartOfPacketIndex = i;
				_asciiOnDeck.runningDataIndexOfStart = _runningDataIndex;
				_asciiOnDeck.timeFound = timestamp;

				asciiStartFoundInProvidedBuffer = true;
			}
			else if (_asciiOnDeck.currentlyBuildingAsciiPacket && data[i] == AsciiEndChar1)
			{
				_asciiOnDeck.asciiEndChar1Found = true;
			}
			else if (_asciiOnDeck.asciiEndChar1Found)
			{
				if (data[i] == AsciiEndChar2)
				{
					// We have a possible data packet.
					size_t runningIndexOfPacketStart = _asciiOnDeck.runningDataIndexOfStart;
					uint8_t* startOfAsciiPacket = NULL;
					size_t packetLength = 0;

					if (asciiStartFoundInProvidedBuffer)
					{
						// All the packet was in this data buffer so we don't
						// need to do any copying.

						startOfAsciiPacket = data + _asciiOnDeck.possibleStartOfPacketIndex;
						packetLength = i - _asciiOnDeck.possibleStartOfPacketIndex + 1;
					}
					else
					{
						// The packet was split between the running data buffer
						// the current data buffer. We need to copy the data
						// over before further processing.

						if (_bufferAppendLocation + i < _bufferSize)
						{
							memcpy(_buffer + _bufferAppendLocation, data, i + 1);

							startOfAsciiPacket = _buffer + _asciiOnDeck.possibleStartOfPacketIndex;
							packetLength = _bufferAppendLocation + i + 1 - _asciiOnDeck.possibleStartOfPacketIndex;
						}
						else
						{
							// We are about to overflow our buffer. Just fall
							// through to reset tracking.
						}
					}

					Packet p(reinterpret_cast<char*>(startOfAsciiPacket), packetLength);

					if (p.isValid())
						dispatchPacket(p, runningIndexOfPacketStart, _asciiOnDeck.timeFound);
				}
				
				// Either this is an invalid packet or was a packet that was processed.
				if (_binaryOnDeck.empty())
					resetTracking();
				else
					_asciiOnDeck.reset();
				asciiStartFoundInProvidedBuffer = false;
			}
			else if (i + 1 > MaximumSizeForAsciiPacket)
			{
				// This must not be a valid ASCII packet.
				if (_binaryOnDeck.empty())
				{
					resetTracking();
				}
				else
				{
					_asciiOnDeck.reset();
					asciiStartFoundInProvidedBuffer = false;
				}
			}

			// Update all of our binary packets on deck.
			queue<BinaryTracker> invalidPackets;
			for (list<BinaryTracker>::iterator it = _binaryOnDeck.begin(); it != _binaryOnDeck.end(); ++it)
			{
				BinaryTracker &ez = (*it);

				if (!ez.groupsPresentFound)
				{
					// This byte must be the groups present.
					ez.groupsPresentFound = true;
					ez.groupsPresent = data[i];
					ez.numOfBytesRemainingToHaveAllGroupFields = 2 * countSetBits(data[i]);

					continue;
				}

				if (ez.numOfBytesRemainingToHaveAllGroupFields != 0)
				{
					// We found another byte belonging to this possible binary packet.
					ez.numOfBytesRemainingToHaveAllGroupFields--;

					if (ez.numOfBytesRemainingToHaveAllGroupFields == 0)
					{
						// We have all of the group fields now.
						size_t remainingBytesForCompletePacket;
						if (ez.startFoundInProvidedDataBuffer)
						{
							size_t headerLength = i - ez.possibleStartIndex + 1;
							remainingBytesForCompletePacket = Packet::computeBinaryPacketLength(reinterpret_cast<char*>(data) + ez.possibleStartIndex) - headerLength;
						}
						else
						{
							// Not all of the packet's group is inside the caller's provided buffer.

							// Temporarily copy the rest of the packet to the receive buffer
							// for computing the size of the packet.

							size_t numOfBytesToCopyIntoReceiveBuffer = i + 1;
							size_t headerLength = _bufferAppendLocation - ez.possibleStartIndex + numOfBytesToCopyIntoReceiveBuffer;

							if (_bufferAppendLocation + numOfBytesToCopyIntoReceiveBuffer < _bufferSize)
							{
								std::memcpy(_buffer + _bufferAppendLocation, data, numOfBytesToCopyIntoReceiveBuffer);

								remainingBytesForCompletePacket = Packet::computeBinaryPacketLength(reinterpret_cast<char*>(_buffer) + ez.possibleStartIndex) - headerLength;
							}
							else
							{
								// About to overrun our receive buffer!
								invalidPackets.push(ez);

								// TODO: Should we just go ahead and clear the ASCII tracker
								//       and buffer append location?

								continue;
							}
						}

						if (remainingBytesForCompletePacket > MaximumSizeExpectedForBinaryPacket)
						{
							// Must be a bad possible binary packet.
							invalidPackets.push(ez);
						}
						else
						{
							ez.numOfBytesRemainingForCompletePacket = remainingBytesForCompletePacket;
						}
					}

					continue;
				}

				// We are currently collecting data for our packet.

				ez.numOfBytesRemainingForCompletePacket--;

				if (ez.numOfBytesRemainingForCompletePacket == 0)
				{
					// We have a possible binary packet!

					uint8_t* packetStart;
					size_t packetLength;

					if (ez.startFoundInProvidedDataBuffer)
					{
						// The binary packet exists completely in the user's provided buffer.
						packetStart = data + ez.possibleStartIndex;
						packetLength = i - ez.possibleStartIndex + 1;
					}
					else
					{
						// The packet is split between our receive buffer and the user's buffer.
						size_t numOfBytesToCopyIntoReceiveBuffer = i + 1;

						if (_bufferAppendLocation + numOfBytesToCopyIntoReceiveBuffer < _bufferSize)
						{
							std::memcpy(_buffer + _bufferAppendLocation, data, numOfBytesToCopyIntoReceiveBuffer);

							packetStart = _buffer + ez.possibleStartIndex;
							packetLength = _bufferAppendLocation - ez.possibleStartIndex + i + 1;
						}
						else
						{
							// About to overrun our receive buffer!
							invalidPackets.push(ez);

							continue;
						}
					}

					Packet p(reinterpret_cast<char*>(packetStart), packetLength);

					if (!p.isValid())
					{
						// Invalid packet!
						invalidPackets.push(ez);
					}
					else
					{
						// We have a valid binary packet!!!.

						// Copy data out of the tracking lists since we will be resetting them.
						BinaryTracker bt = ez;

						invalidPackets = queue<BinaryTracker>();
						resetTracking();

						dispatchPacket(p, bt.runningDataIndexOfStart, bt.timeFound);

						break;
					}
				}
			}

			// Remove any invalid packets.
			while (!invalidPackets.empty())
			{
				_binaryOnDeck.remove(invalidPackets.front());
				invalidPackets.pop();
			}

			if (_binaryOnDeck.empty() && !_asciiOnDeck.currentlyBuildingAsciiPacket)
			{
				_bufferAppendLocation = 0;
			}

			if (data[i] == BinaryStartChar)
			{
				// Possible start of a binary packet.
				_binaryOnDeck.push_back(BinaryTracker(i, _runningDataIndex, timestamp));
			}
		}

		if (_binaryOnDeck.empty() && !_asciiOnDeck.currentlyBuildingAsciiPacket)
			// No data to copy over.
			return;

		// Perform any data copying to our receive buffer.

		size_t dataIndexToStartCopyingFrom = 0;
		bool binaryDataToCopyOver = false;
		size_t binaryDataMoveOverIndexAdjustment = 0;

		if (!_binaryOnDeck.empty())
		{
			binaryDataToCopyOver = true;

			if (_binaryOnDeck.front().startFoundInProvidedDataBuffer)
			{
				dataIndexToStartCopyingFrom = _binaryOnDeck.front().possibleStartIndex;
				binaryDataMoveOverIndexAdjustment = dataIndexToStartCopyingFrom;
			}
		}

		if (_asciiOnDeck.currentlyBuildingAsciiPacket && asciiStartFoundInProvidedBuffer)
		{
			if (_asciiOnDeck.possibleStartOfPacketIndex < dataIndexToStartCopyingFrom)
			{
				binaryDataMoveOverIndexAdjustment -= binaryDataMoveOverIndexAdjustment - _asciiOnDeck.possibleStartOfPacketIndex;
				dataIndexToStartCopyingFrom = _asciiOnDeck.possibleStartOfPacketIndex;
			}
			else if (!binaryDataToCopyOver)
			{
				dataIndexToStartCopyingFrom = _asciiOnDeck.possibleStartOfPacketIndex;
			}

			// Adjust our ASCII index to be based on the recieve buffer.
			_asciiOnDeck.possibleStartOfPacketIndex = _bufferAppendLocation + _asciiOnDeck.possibleStartOfPacketIndex - dataIndexToStartCopyingFrom;
		}

		// Adjust any binary packet indexes we are currently building.
		for (list<BinaryTracker>::iterator it = _binaryOnDeck.begin(); it != _binaryOnDeck.end(); ++it)
		{
			if ((*it).startFoundInProvidedDataBuffer)
			{
				(*it).startFoundInProvidedDataBuffer = false;
				(*it).possibleStartIndex = (*it).possibleStartIndex - binaryDataMoveOverIndexAdjustment + _bufferAppendLocation;
			}
		}

		if (_bufferAppendLocation + length - dataIndexToStartCopyingFrom < _bufferSize)
		{
			// Safe to copy over the data.

			size_t numOfBytesToCopyOver = length - dataIndexToStartCopyingFrom;
			uint8_t *copyFromStart = data + dataIndexToStartCopyingFrom;

			std::memcpy(_buffer + _bufferAppendLocation, copyFromStart, numOfBytesToCopyOver);
			_bufferAppendLocation += numOfBytesToCopyOver;
		}
		else
		{
			// We are about to overflow our buffer.
			resetTracking();
		}
	}

	void dispatchPacket(Packet &packet, size_t runningDataIndexAtPacketStart, TimeStamp timestamp)
	{
		if (_possiblePacketFoundHandler != NULL)
		{
			_possiblePacketFoundHandler(_possiblePacketFoundUserData, packet, runningDataIndexAtPacketStart, timestamp);
		}
	}
};

#if defined(_MSC_VER) && _MSC_VER <= 1600
	#pragma warning(push)
	// Disable VS2010 warning for 'this' used in base member initializer list.
	#pragma warning(disable:4355)
#endif

PacketFinder::PacketFinder() :
	_pi(new Impl(this))
{
}

PacketFinder::PacketFinder(size_t internalReceiveBufferSize) :
	_pi(new Impl(this, internalReceiveBufferSize))
{
}

#if defined(_MSC_VER) && _MSC_VER <= 1600
	#pragma warning(pop)
#endif

PacketFinder::~PacketFinder()
{
	delete _pi;
}

void PacketFinder::processReceivedData(char data[], size_t length)
{
	TimeStamp placeholder;

	processReceivedData(data, length, placeholder);
}

void PacketFinder::processReceivedData(char data[], size_t length, TimeStamp timestamp)
{
	_pi->dataReceived(reinterpret_cast<uint8_t*>(data), length, timestamp);
}

#if PYTHON

void PacketFinder::processReceivedData(boost::python::list data)
{
	//size_t len = boost::python::len(data);
	//unique<char[]> cppData = make_unique<char[]>(len);

	//for (auto i = 0; i < len; i++)
	//	cppData[i] = boost::python::extract<int>(data[i]);

	//processReceivedData(cppData.get(), len);
}

#endif

void PacketFinder::registerPossiblePacketFoundHandler(void* userData, ValidPacketFoundHandler handler)
{
	if (_pi->_possiblePacketFoundHandler != NULL)
		throw invalid_operation();

	_pi->_possiblePacketFoundHandler = handler;
	_pi->_possiblePacketFoundUserData = userData;
}

void PacketFinder::unregisterPossiblePacketFoundHandler()
{
	if (_pi->_possiblePacketFoundHandler == NULL)
		throw invalid_operation();

	_pi->_possiblePacketFoundHandler = NULL;
	_pi->_possiblePacketFoundUserData = NULL;
}

#if PYTHON

//void PacketFinder::register_packet_found_handler(boost::python::object* callable)
boost::python::object* PacketFinder::register_packet_found_handler(PyObject* callable/*boost::python::object* callable*/)
{
	_pi->_pythonPacketFoundHandler = callable;
	callable->ob_refcnt++;
	//_pi->_pythonPacketFoundHandler = callable;
	//callable->ptr()->ob_refcnt++;
	//callable->ptr();
	//_pi->_rawDataReceivedHandlerPython = callable;
	//callable->ob_refcnt++;

	return NULL;

	//return Py_None;
}

#endif

}
}
}
